mautic/mautic
# separate local config from application config

| Q                                      | A
| -------------------------------------- | ---
| Bug fix? (use the a.b branch)          | [ ]
| New feature/enhancement? (use the a.x branch)      | [x]
| Deprecations?                          | [ ]
| BC breaks? (use the c.x branch)        | [ ]
| Automated tests included?              | [ ] 
| Related user documentation PR URL      | mautic/mautic-documentation#... 
| Related developer documentation PR URL | mautic/developer-documentation#... 
| Issue(s) addressed                     | Fixes #... 



#### Description:

This PR separates the 2 types of configuration that exist in Mautic:
* **application configuration** that belong to the Mautic, stored in `<web_root>/app/config`
* **user configuration** that contains specific config for the Mautic installation (e.g. db credentials, mailer config, ...) and overrides of other application config (e.g. paths), stored in `<project_root>/local_config`

This ensures that setups that use composer don't loose the config on an update.

for non-composer setups, `<project_root>` and  `<web_root>` are the same, but for composer setups, the `<web_root>` is 1 level deeper.

In general it's a good idea to separate the application from user configuration. Thus was already done for media and assets in https://github.com/mautic/mautic/pull/10682



#### Steps to test this PR:


1. Install Mautic 5.x branch locally, without applying this PR
2. change some mautic config via the UI, and save it.
3. override some config via the `app/config/parameters_local.php` file
4. apply the patch, and run the migration
5. verify all local config files are no longer present in `<web_root>/app/config`
6. verify all local config files are no present in `<project_root>/local_config`
7. verify your Mautic installation is still working
